### DISCUSSION TITLE
Authentication in Message Extensions

### DISCUSSION THREAD

**Discussion by garrytrinder** [microsoft/teams-sdk#438] [Q&A] [answered]
I have the following application configuration.

```ts
const app = new Application<ApplicationTurnState>({
    adapter,
    botAppId: config.aadAppId,
    storage,    
    authentication: {
        connectionName: "MicrosoftGraph",
        title: "Sign In"
    }
});
```

This enables me to retrieve an access token from state.

```ts
const { authToken } = state.temp.value;
```

However this currently only works for a bot, how can I get this to work in a message extension? Is it supported?

Nothing appears to be exposed in `state` or `context.activity` in the query handler.

**Comment by garrytrinder** [comment]
I've managed to get something working, but my workaround involves creating and wiring up an activity handler to override the `handleTeamsMessagingExtensionQuery` function. 

I'm using TeamsFx to simplify the authentication flow, using the `handleMessageExtensionQueryWithSSO` which returns a `MessageExtensionResponse` to start the auth flow. I tried to use this function inside the query handler, but `app.messageExtensions.query` expects a `MessageExtensionResult` response instead, so I couldn't get them to play nicely together. /cc @therealjohn

**Reply by corinagum** [reply]
Hi @garrytrinder, thanks for your patience! 
here is the[ documentation on SSO for Message Extensions](https://learn.microsoft.com/en-us/microsoftteams/platform/bots/how-to/authentication/bot-sso-overview)

Right now, this is how you would need to do it even if using teams-ai sdk.

**Reply by garrytrinder** [reply]
Thanks for the reply @corinagum I've used the example provided by @blackchoey and I am unlocked.

**Comment by blackchoey** [comment, marked as answer]
@garrytrinder I created an example to use `handleMessageExtensionQueryWithSSO` from `@microsoft/TeamsFx` package to implement SSO: https://github.com/blackchoey/message-extension-sso and solved the type mismatch problem. You can have a try.
This is not the final solution in Teams-AI library. Please stay tuned for the updates of the library.

**Reply by garrytrinder** [reply]
Perfect! Thanks @blackchoey

Are there plans to move this into Teams AI library?

**Reply by blackchoey** [reply]
We're discussing on this. Still need some time to figure out the experience in Teams AI library. Please stay tuned.
